lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -4499,6 +4569,203 @@ void MySQL_HostGroups_Manager::read_only_action(char *hostname, int port, int re
+ * 
+ */
+void MySQL_HostGroups_Manager::read_only_action_v2(const std::list<std::tuple<std::string,int,int>>& mysql_servers) {
+
+	std::string hostname;
+	int port = -1;
+	int read_only = -1;
+	bool update_mysql_servers_table = false;
+
+	unsigned long long curtime1 = monotonic_time();
+	wrlock();
+	for (const auto& server : mysql_servers) {
+		bool is_writer = false;
+		std::tie(hostname, port, read_only) = server;
+		const std::string& srv_id = hostname + ":::" + std::to_string(port);
+		
+		auto itr = hostgroup_server_mapping.find(srv_id);
+
+		if (itr == hostgroup_server_mapping.end()) {
+			proxy_warning("Server %s:%d not found\n", hostname.c_str(), port);
+			continue;
+		}
+
+		HostGroup_Server_Mapping* host_server_mapping = itr->second.get();
+
+		if (!host_server_mapping)
+			assert(0);
+
+		const std::vector<HostGroup_Server_Mapping::Node>& writer_map = host_server_mapping->get(HostGroup_Server_Mapping::Type::WRITER);
+
+		is_writer = !writer_map.empty();
+
+		if (read_only == 0) {
+			if (is_writer == false) {
+				// the server has read_only=0 (writer), but we can't find any writer, 
+				// so we copy all reader nodes to writer
+				host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::WRITER, HostGroup_Server_Mapping::Type::READER);
+
+				if (mysql_thread___monitor_writer_is_also_reader) {
+					// server is also a reader, we copy all nodes from writer to reader (previous reader nodes will be reused)
+					host_server_mapping->copy_if_not_exists(HostGroup_Server_Mapping::Type::READER, HostGroup_Server_Mapping::Type::WRITER);
+				} else {
+					// server can only be a writer
+					host_server_mapping->clear(HostGroup_Server_Mapping::Type::READER);
+				}
+
+				update_mysql_servers_table = true;
+			} else {
+				bool act = false;
+
+				// if the server was RO=0 on the previous check then no action is needed
+				if (host_server_mapping->get_readonly_flag() != 0) {
+					// it is the first time that we detect RO on this server
+					const std::vector<HostGroup_Server_Mapping::Node>& reader_map = host_server_mapping->get(HostGroup_Server_Mapping::Type::READER);
+
+					for (const auto& reader_node : reader_map) {
+						for (const auto& writer_node : writer_map) {
+
+							if (reader_node.writer_hostgroup_id == writer_node.writer_hostgroup_id) {
+								goto __writer_found;
+							}
+						}
+						act = true;
+						break;
+					__writer_found:
+						continue;
+					}
+
+					if (act == false) {
+						// no action required, therefore we set readonly_flag to 0
+						proxy_info("read_only_action_v2() detected RO=0 on server %s:%d for the first time after commit(), but no need to reconfigure\n", hostname.c_str(), port);
+						host_server_mapping->set_readonly_flag(0);
+					}
+				} else {
+					// the server was already detected as RO=0
+					// no action required
+				}
+
+				if (act == true) {	// there are servers either missing, or with stats=OFFLINE_HARD
+
```

Comment:
Following the same logic as before, here I propose something like:

```
proxy_info("Server '%s:%d' found missing from 'writer_hostgroup'\n", hostname.c_str(), port);
```

Reply:
Done

Reply:
I see that that the proposed logging here has been moved above to lines `4637/4638`, issue with this is that message will be logged for each `read_only_action` call once the servers are in the destination hostgroups, causing excessive output in the error log:

```
2023-03-14 12:16:44 MySQL_HostGroups_Manager.cpp:4638:read_only_action_v2(): [INFO] Server '127.1.1.14:3306' found with 'read_only=0', also found as writer
2023-03-14 12:16:44 MySQL_HostGroups_Manager.cpp:4638:read_only_action_v2(): [INFO] Server '127.1.1.12:3306' found with 'read_only=0', also found as writer
```

Since I'm reading this again, thinking about the previously proposed message, I would slightly change the wording into:

```
proxy_info("Server '%s:%d' with 'read_only=0' found missing at some 'writer_hostgroup'\n", hostname.c_str(), port);
```

This is to make clear the point that the server was found as writer, but still was found missing at some writer hostgroup, and that's the context for actions to be taken.

Reply:
Done